### ISSUE TITLE
Using scatter for sparse max/min

### ISSUE THREAD

**Issue by smorad** [rusty1s/pytorch_scatter#257]
How would I use scatter to aggregate across an arbitrary dimension of a sparse tensor using max, min, etc?

Let's say I have some 4D indices and I want the `max` across the `-2` dim:

```python
import torch
import torch_scatter


idx = torch.tensor([
        [0, 0, 0, 0, 0],
        [0, 0, 0, 1, 1],
        [1, 2, 2, 1, 2],
        [0, 0, 1, 0, 1]
])

values = torch.tensor([1, 2, 3, 4, 5])

# This is what I want returned, the max over dim=-2
results = torch.tensor([1, 3, 4, 5])
```

One way to do this would be to flatten `idx` into a 1D tensor, and offset each dim, e.g.
```python
idx_1d = idx[0] * idx[1].max() + idx[1]
torch_scatter.scatter(idx_1d, values, reduce="max")
```
But this does not expand nicely to `n` indexing dimensions. Furthermore, for many dimensions the product of `idx[1].max() * idx[2].max(), ...` could overflow.

Is there a better way to use `scatter` to take the max/mean/min/etc of a sparse matrix?

**Comment by rusty1s** [comment]
I'm afraid I'm not totally sure what you are trying to do as the dimensions of `idx` and `values` do not match. If you want to aggregate in the `-2` dim, the following one works for me:
```python
idx = torch.tensor([
    [0, 0, 0, 0, 0],
    [0, 0, 0, 1, 1],
    [1, 2, 2, 1, 2],
    [0, 0, 1, 0, 1],
])

values = torch.tensor([1, 2, 3, 4, 5])
values = values.view(1, -1).repeat(4, 1)

out = torch_scatter.scatter(idx, values, dim=-2, reduce="max")
```

**Comment by smorad** [comment]
Each K-dimensional value is associated with a 4D index. So if two 4D indices are not unique (e.g. `0 1 0 1` and `0 1 0 1`), they will map to a single output in a 1xK-dimensional vector. Conversely, `0 1 0 1` and `0 1 0 0` will map to different entries in a 1D vector.

Essentially, scatter becomes a reduce operation. This lets us perform operations that you have implemented for sparse tensors, that are not currently supported on vanilla sparse torch tensors (like `max`, `min`, `logsumexp`, etc).

I currently flatten the 4D indices into unique 1D indices by computing "strides"
```python
def flatten_idx_n_dim(idx):
    assert idx.ndim == 2
    strides = idx.max(dim=1).values + 1
    new_idx = torch.zeros(idx.shape[-1], dtype=torch.long, device=idx.device)
    offsets = []

    for i in range(len(strides) - 1):
        offset = strides[i + 1:].prod()
        offsets.append(offset)
        new_idx += offset * idx[i]

    new_idx += idx[-1]

    return new_idx, offsets
```

So that I can take the max of a sparse tensor
```python
flat_scat_idx, offsets = flatten_idx_n_dim(n_dimensional_idx)
maxes, argmax = scatter_max(values, flat_scat_idx) # Take the max over a sparse tensor, which is not yet possible using torch
```

Is there a better way to do this using `torch_sparse`?

**Comment by rusty1s** [comment]
I think flattening is the way to go here. The problem with your approach is that the flattened indices can become quite large, while the effective number of indices is quite small. You can map indices to a contiguous range via:
```python
idx, _ = flatten_idx_n_dim(idx)
print(idx)
_, idx = torch.unique(idx, return_inverse=True)
print(idx)
```
